src/lib.rs

```diff
@@ -1,7 +1,552 @@
+                let c_1 = c - c_0;
+
+                // since we don't control `c_1` the response for these proofs will be legitimate
+                res_G_1s[i] = s!({ &res_G_1s[i] } + { c_1.into_secp256k1() } * { &r_is[i] })
+                    .mark::<NonZero>()
+                    .expect("non-zero response");
+                res_H_1s[i] += c_1.into_ed25519() * s_is[i];
+
+                c_0s.push(c_0);
+                c_1s.push(c_1);
+            } else {
+                let c_1 = c_cheats[i];
+                // compute challenge for the actual bit
+                let c_0 = c - c_1;
+
+                // since we don't control `c_0` the response for these proofs will be legitimate
+                res_G_0s[i] = s!({ &res_G_0s[i] } + { c_0.into_secp256k1() } * { &r_is[i] })
+                    .mark::<NonZero>()
+                    .expect("non-zero response");
+                res_H_0s[i] += c_0.into_ed25519() * s_is[i];
+
+                c_0s.push(c_0);
+                c_1s.push(c_1);
+            };
+        }
+
+        // sum of `r_i * 2^i` for all `i`
+        let r = secp256k1::blinder_sum(&r_is);
+        // sum of `s_i * 2^i` for all `i`
+        let s = ed25519::blinder_sum(s_is);
+
+        Proof {
+            C_G_is,
+            C_H_is,
+            c_0s,
+            c_1s,
+            U_G_0s,
+            U_H_0s,
+            U_G_1s,
+            U_H_1s,
+            res_G_0s,
+            res_H_0s,
+            res_G_1s,
+            res_H_1s,
+            r,
+            s,
+        }
+    }
+
+    /// Compute challenge by hashing the statements (Pedersen Commitments) and
+    /// announcements for each bitwise proof.
+    pub fn compute_challenge(
+        C_G_is: &[secp256k1::Point],
+        C_H_is: &[ed25519::Point],
+        U_G_0s: &[secp256k1::Point],
+        U_G_1s: &[secp256k1::Point],
+        U_H_0s: &[ed25519::Point],
+        U_H_1s: &[ed25519::Point],
+    ) -> Scalar {
+        let mut hasher = Sha256::default();
+
+        for (i, (C_G_i, C_H_i)) in C_G_is.iter().zip(C_H_is.iter()).enumerate() {
+            hasher.update(C_G_i.to_bytes());
+            hasher.update(C_H_i.compress().as_bytes());
+            hasher.update(U_G_0s[i].to_bytes());
+            hasher.update(U_G_1s[i].to_bytes());
+            hasher.update(U_H_0s[i].compress().as_bytes());
+            hasher.update(U_H_1s[i].compress().as_bytes());
+        }
+
+        let hash = hasher.finalize();
+
+        let mut bytes = [0u8; 32];
+        bytes.copy_from_slice(hash.as_slice());
+
+        Scalar::from(bytes)
+    }
+
+    /// Verify the proof.
+    ///
+    /// In particular:
+    ///
+    /// 1. Ensure that the combinations of Pedersen Commitments actually
+    /// represent the public values `xG` and `xH`.
+    /// 2. Ensure that each bitwise response satisfies its corresponding
+    /// challenge.
+    pub fn verify(&self, xG: &secp256k1::Point<Jacobian>, xH: ed25519::Point) -> Result<(), Error> {
```

Comment:
It's preferable to be `impl PointType` so it takes any point. `Jacobian` is a weirdly particular thing to require the caller to give you.

Reply:
Thanks for the tip! I wanted to ensure that the point passed was Jacobian since the docs point out that it is most efficient to work with them. But you are right that the caller should be free to pass any point.